### PR TITLE
[AOTI] Avoid serializing full storage for shared-storage constants

### LINKED ISSUE #186176: AOTInductor serializes entire underlying storage for each constant tensor, causing severe bloat for shared-storage weights
### 🐛 Describe the bug

AOTInductor's AOTI binary constant serialization appears to serialize each constant tensor's entire underlying storage, rather than the tensor's logical payload.

This causes severe package size bloat when multiple constant tensors share the same underlying storage.

A common way to hit this is `nn.RNN.flatten_parameters()`, which packs RNN weights into one contiguous flat storage. Each individual RNN parameter becomes a view into that flat storage with its own `storage_offset`.

In PyTorch 2.12.0, the AOTI binary serialization path in `torch/_inductor/codecache.py` still does:

```python
t_cpu = t.untyped_storage().cpu()
data_ptr = t_cpu.data_ptr()
nbytes = t_cpu.nbytes()
````

inside `_to_bytes()`.

So if 8 RNN weight tensors share the same ~194 MB flat storage, AOTI serializes the full ~194 MB storage 8 times.

---

### Expected behavior

AOTI should not serialize the same shared storage once per tensor.

Possible correct behaviors:

1. serialize each tensor's logical payload only, respecting `storage_offset`, `size`, `stride`, and `dtype`; or
2. serialize each unique storage once and record tensor metadata such as `storage_id`, `storage_offset`, `sizes`, and `strides`.

---

### Actual behavior

For RNN flattened weights, `binary_blob` serializes the entire shared storage for each constant tensor.

In my repro, the model has 8 RNN parameters sharing one flat storage.

Actual parameter bytes:

```text
Actual param bytes: 203,489,280 = 194.1 MB
Unique storage:     203,489,280 = 194.1 MB
```

But AOTI `binary_blob` output size is:

```text
binary_blob shared Storage: 1553.1 MB
binary_blob cloned Storage: 194.7 MB
pickle_weights shared:     194.7 MB

binary_blob bloat ratio: 8.0x
```

This is exactly `8 * ~194 MB`, matching the number of shared-storage RNN parameters.

---

### Repro details

Model configuration:

```python
import torch
import torch.nn as nn

model = nn.RNN(
    input_size=128,
    hidden_size=4096,
    num_layers=2,
).cuda().eval()

model.flatten_parameters()
```

After `flatten_parameters()`, the RNN parameters share one underlying storage.

Before workaround, each parameter reports a much larger storage size than its logical tensor size:

```python
storage_nbytes = p.untyped_storage().nbytes()
tensor_nbytes = p.numel() * p.element_size()
```

Workaround:

```python
def clone_shared_weight_storages(model: torch.nn.Module) -> int:
    from collections import defaultdict

    storage_to_params = defaultdict(list)

    for name, p in model.named_parameters():
        if p.numel() == 0:
            continue

        storage = p.untyped_storage()
        key = (
            p.device.type,
            p.device.index,
            storage.data_ptr(),
            storage.nbytes(),
        )
        storage_to_params[key].append((name, p))

    cloned = 0

    for _, params in storage_to_params.items():
        shared_storage = len(params) > 1

        for name, p in params:
            storage_nbytes = p.untyped_storage().nbytes()
            tensor_nbytes = p.numel() * p.element_size()

            oversized_view = (
                p.storage_offset() != 0
                or storage_nbytes != tensor_nbytes
            )

            if not shared_storage and not oversized_view:
                continue

            cloned_tensor = p.detach().clone(memory_format=torch.contiguous_format)
            cloned_param = torch.nn.Parameter(
                cloned_tensor,
                requires_grad=p.requires_grad,
            )

            parent_path, _, leaf_name = name.rpartition(".")
            parent = model.get_submodule(parent_path) if parent_path else model
            setattr(parent, leaf_name, cloned_param)

            cloned += 1

    return cloned
```

After cloning, each parameter owns compact storage:

```text
Name          Shape          Actual(B)    Storage(B)   Offset   Ratio
weight_ih_l0  [4096, 128]      2,097,152    2,097,152       0    1.0x
weight_hh_l0  [4096, 4096]    67,108,864   67,108,864       0    1.0x
bias_ih_l0    [4096]              16,384       16,384       0    1.0x
bias_hh_l0    [4096]              16,384       16,384       0    1.0x
weight_ih_l1  [4096, 4096]    67,108,864   67,108,864       0    1.0x
weight_hh_l1  [4096, 4096]    67,108,864   67,108,864       0    1.0x
bias_ih_l1    [4096]              16,384       16,384       0    1.0x
bias_hh_l1    [4096]              16,384       16,384       0    1.0x
```

The AOTI `binary_blob` package size then drops from 1553.1 MB to 194.7 MB.

---

### Affected configuration

This affects the AOTI binary constant serialization path:

```python
config.aot_inductor.package_constants_in_so
or config.aot_inductor.package_constants_on_disk_format == "binary_blob"
```

In PyTorch 2.12.0, `package_constants_in_so` defaults to `True`.

---

### Environment

```text
PyTorch version: 2.12.0+cu130
CUDA used to build PyTorch: 13.0
CUDA runtime version: 13.0.88
GPU: NVIDIA A100-SXM4-40GB
Nvidia driver version: 550.127.08
cuDNN: 9.15.0 / 9.20.0.48
Python: 3.12.3
OS: Ubuntu 24.04.3 LTS
GCC: 13.3.0
Component: AOTInductor / AOTI
Path: torch/_inductor/codecache.py
```
test file
[repro_aoti_shared_storage_bloat.py](https://github.com/user-attachments/files/28578348/repro_aoti_shared_storage_bloat.py)

### Versions

PyTorch version: 2.12.0+cu130
CUDA used to build PyTorch: 13.0
OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: 13.3.0
CMake version: 3.31.6
Python version: 3.12.3
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to: LAZY
GPU: NVIDIA A100-SXM4-40GB
Nvidia driver version: 550.127.08
cuDNN version: 9.15.0 / 9.20.0.48 from installed packages
CPU: AMD EPYC 7742 64-Core Processor
pip torch: torch==2.12.0
pip torchvision: torchvision==0.27.0
pip pytorch-triton: pytorch-triton==3.5.0+gitde3506d2
conda: Could not collect

cc @mruberry @mikaylagawarecki @chauhang @penguinwu @avikchaudhuri @zhxchen17 @tugsbayasgalan @angelayi @ydwu4 @desertfire @yushangdi @jataylo @iupaikov-amd

### PR DESCRIPTION
**Pull request by wasimat404**
Fixes #186176

AOTI's binary-blob constant serialization wrote each constant's entire
`untyped_storage()`. When constants share one storage (e.g. RNN params
after `flatten_parameters()`), every view serialized the whole buffer,
bloating the blob to ~N× for N shared views.

For contiguous constants, this computes the serialized size from the
tensor's logical extent (`(storage_offset + numel) * element_size`)
instead of the full storage, applied consistently in the three places
that must agree: `_constant_nbytes` and the `_worker` copy in
`codecache.py`, and `data_size` in `cpp_wrapper_cpu.py`. The reader
recovers tensors via `storage_offset` within the slice, so `offset` is
unchanged and **no runtime reader changes are needed**. Non-contiguous
and mkldnn constants keep the previous behavior.

This stays within the existing reader contract, so it removes the N×
duplication but isn't maximally compact (a large-offset constant still
includes leading padding). Full per-constant compaction would need a
reader-contract change — separate follow-up.

Added `test_shared_storage_blob_not_bloated` (nn.RNN via `binary_blob`).

> and run the test. Happy to iterate on the bound or approach per CI +
> review.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude explain those CI failures, check if they are related, and review the changes in this PR

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by eellison** [comment]
deferring aoti to @shangdi @desertfire

**Review by desertfire — CHANGES_REQUESTED** [review]
I don't think this is the right way to solve the problem. We should use storage pointer to dedup.

**Comment by wasimat404** [comment]
> I don't think this is the right way to solve the problem. We should use storage pointer to dedup.

Agreed..but mine trims each constant separately, so shared storage still gets written once per tensor. I'll group by storage pointer, write each one once, and update the loader so shared constants point to the same chunk. Where would you prefer the per-constant "which storage" mapping to live in the metadata?
